### ISSUE TITLE
GPT3.5 continuous conversations?

### ISSUE THREAD

**Issue by centminmod** [enricoros/nextjs-chatgpt-app#76]
Great work @enricoros on latest updates! I'm trying to understand why the app and OpenAI AI token limit utilisation works the way it does for your app. When you hit 4097 token limit we need to remove previous messages. But why doesn't https://chat.openai.com/chat implementation of GPT3.5 don't need to do so? Is their method of implementation possible for your app?

![image](https://user-images.githubusercontent.com/5899584/231278216-90b5206b-eac5-47e3-9a00-da061bd7d672.png)

**Comment by enricoros** [comment]
Do you know what's their method of implementation? That's a very interesting concept.

Possibly the chat that's submitted includes the system message, and then walks back from the last message up to when the context window is full. I could actually do this.

Let me know if you can reverse engineer how they do it, before I take a stab.

**Comment by centminmod** [comment]
Yeah I'm unfamiliar with how OpenAI's ChatGPT 3.5 works so can't help there. But I asked ChatGPT Plus GPT4 the question and it gave me this :)

Q. How does https://chat.openai.com/chat GPT3.5 allow continous chat conversations even after reaching 4097 token limit while other app implementations using OpenAI API hit a limit at 4097 tokens and require deleting previous messages first

A. The OpenAI API has a 4096 token limit per request, which means that any message longer than 4096 tokens will need to be broken up into multiple requests. However, the web-based chat interface at https://chat.openai.com/chat uses a technique called "streaming" to allow for continuous conversations that exceed the 4096 token limit.

Streaming involves breaking up the conversation into smaller "chunks" and sending them to the API one at a time, while keeping the connection between the client (i.e. the web browser) and the server (i.e. the OpenAI API) open. As new chunks are received from the server, they are displayed in the chat interface in real-time, allowing for a seamless conversation experience.

In the case of the OpenAI chat interface, each chunk is limited to 2048 tokens, which is half of the API limit. When a user types a message that exceeds 2048 tokens, it is broken up into multiple chunks and sent to the server one chunk at a time. The server processes each chunk and sends back a response, which is displayed in the chat interface. This process continues until the conversation is complete.

By using this streaming approach, the OpenAI chat interface is able to provide a seamless conversation experience, even for long conversations that exceed the 4096 token limit. However, it's worth noting that this approach requires a persistent connection between the client and server, which may not be feasible for all applications depending on the resources available.

Q. Can you provide more technical details of OpenAI Streaming method

A. Sure, here are some additional technical details about how OpenAI's streaming method works:

When a user types a message in the chat interface, the message is initially stored in a local buffer in the user's web browser. The buffer is then split into smaller "chunks" of text, with each chunk containing up to 2048 tokens.

The first chunk is then sent to the OpenAI API using a WebSocket connection. A WebSocket is a protocol for creating a persistent, bidirectional communication channel between a client and a server. The WebSocket allows the client (i.e. the user's web browser) to send data to the server (i.e. the OpenAI API) and receive data from the server in real-time.

When the OpenAI API receives the first chunk of the user's message, it processes it and sends a response back to the client over the same WebSocket connection. The response contains the first part of the AI's generated text, based on the input chunk.

The client then displays the response in the chat interface and sends the next chunk of the user's message to the API using the same WebSocket connection. This process repeats until the entire user message has been sent to the API and the AI has generated a complete response.

Throughout the conversation, the WebSocket connection between the client and the OpenAI API remains open, allowing for real-time communication without the need for multiple API requests. This approach minimizes the latency and overhead associated with traditional REST API requests and allows for a more seamless user experience.

It's worth noting that the use of WebSockets and streaming requires additional server-side infrastructure to handle the persistent connections and manage the flow of data. However, for applications that require real-time communication and support for long messages, this approach can provide significant benefits in terms of performance and user experience.

**Comment by enricoros** [comment]
Thanks for digging. I believe this may not be fully correct because it's missing streaming with chunking. I'll inspect the network packets :)

**Comment by enricoros** [comment]
Thanks for looking into this @typpo. I will probably have it as an option, or figure out (e.g. graying out) how to make it explicit for the user.

**Comment by enricoros** [comment]
Just had an (obvious) epiphany @typpo @centminmod.
This should not be a sliding window, what is filling the context should be the result of an attention operation.
Maybe a way you can think about it is by recalling the pinecone feature #9 , but better.

Assume we have a query ready to be submitted by the user, the messages in history should be prioritized by their dot product with the embeddings of the user message - which means that the context will be filled up with things are most relevant first. This will allow, for instance, to use earlier parts of the conversation and remember earlier parts of the conversation if the user implies that.

**Comment by centminmod** [comment]
@enricoros just found https://www.phind.com/ it has a unique way of consuming longer token/character prompts, 1st prompt input takes 3000 characters max and then split it at 2900 chaacters using my tool at https://slicer.centminmod.com/ and then there is a 2nd input form for additional code and i loaded the rest of split text into it and it managed to consume my script and provide a summary. Not sure how it works, but another idea :)

**Comment by enricoros** [comment]
Thanks. We have actually an impressive feature in the codebase, made by @fredliubojin that is only turned on with PDFs right now. It keeps using gpt to compress useful information in the prompt.
May that be the way that phind is doing it?

**Comment by centminmod** [comment]
compressed info sounds interesting. No idea what Phind is doing heh

**Comment by enricoros** [comment]
Closing as obsolete.

Answering the original question: ChatGPT was running a sliding window (thanks @typpo for digging up the measurement), and the 4k era it belonged to is over.

What exists today: the composer shows live tokens - history, your draft, and the reserved response - against the model's context limit, with cost and an excess indicator; and any message can be dropped from what is sent with Skip message in the message menu (also on multi-select), which removes it from the request and from the token count.

What I said here and did not deliver: a sliding-window option with explicit UI. That, and the relevance-ranked stuffing I sketched later in this thread (fill the window by embedding similarity to the pending message, not by recency), now live in #216, which is the single issue for what goes into the request. Closing this one.

@centminmod this close comes from an automated backlog pass - if you disagree, feel free to reopen.
